atom/browser/api/atom_api_notification.cc

```diff
@@ -0,0 +1,199 @@
+#include "atom/common/node_includes.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/threading/thread_task_runner_handle.h"
+#include "native_mate/constructor.h"
+#include "native_mate/dictionary.h"
+#include "third_party/skia/include/core/SkBitmap.h"
+#include "ui/gfx/codec/png_codec.h"
+#include "ui/gfx/image/image.h"
+#include "url/gurl.h"
+
+namespace atom {
+
+namespace api {
+
+int id_counter = 1;
+std::map<int, Notification*> notifications_;
+
+Notification::Notification(v8::Isolate* isolate,
+                           v8::Local<v8::Object> wrapper,
+                           mate::Arguments* args) {
+  InitWith(isolate, wrapper);
+
+  mate::Dictionary opts;
+  if (args->GetNext(&opts)) {
+    opts.Get("title", &title_);
+    opts.Get("body", &body_);
+    has_icon_ = opts.Get("icon", &icon_);
+    if (has_icon_) {
+      opts.Get("icon", &icon_path_);
+    }
+    opts.Get("silent", &silent_);
+    opts.Get("replyPlaceholder", &reply_placeholder_);
+    opts.Get("hasReply", &has_reply_);
+    id_ = id_counter++;
+  }
+  notifications_[id_] = this;
+  OnInitialProps();
+}
+
+Notification::~Notification() {}
+
+// static
+mate::WrappableBase* Notification::New(mate::Arguments* args) {
+  if (!Browser::Get()->is_ready()) {
+    args->ThrowError("Cannot create Notification before app is ready");
+    return nullptr;
+  }
+  return new Notification(args->isolate(), args->GetThis(), args);
+}
+
+bool Notification::HasID(int id) {
+  return notifications_.find(id) != notifications_.end();
+}
+
+Notification* Notification::FromID(int id) {
+  return notifications_[id];
+}
+
+// Getters
+int Notification::GetID() {
+  return id_;
+}
+base::string16 Notification::GetTitle() {
+  return title_;
+}
+base::string16 Notification::GetBody() {
+  return body_;
+}
+bool Notification::GetSilent() {
+  return silent_;
+}
+base::string16 Notification::GetReplyPlaceholder() {
+  return reply_placeholder_;
+}
+bool Notification::GetHasReply() {
+  return has_reply_;
+}
+
+// Setters
+void Notification::SetTitle(base::string16 new_title) {
+  title_ = new_title;
+  NotifyPropsUpdated();
+}
+void Notification::SetBody(base::string16 new_body) {
+  body_ = new_body;
+  NotifyPropsUpdated();
+}
+void Notification::SetSilent(bool new_silent) {
+  silent_ = new_silent;
+  NotifyPropsUpdated();
+}
+void Notification::SetReplyPlaceholder(base::string16 new_reply_placeholder) {
+  reply_placeholder_ = new_reply_placeholder;
+  NotifyPropsUpdated();
+}
+void Notification::SetHasReply(bool new_has_reply) {
+  has_reply_ = new_has_reply;
+  NotifyPropsUpdated();
+}
+
+void Notification::OnClicked() {
+  Emit("click");
+}
+
+void Notification::OnReplied(std::string reply) {
+  Emit("reply", reply);
+}
+
+void Notification::OnShown() {
+  Emit("show");
+}
+
+void Notification::OnClosed() {
+  Emit("close");
+}
+
+void Notification::NotifyPropsUpdated() {}
+
+// Showing notifications
+void Notification::Show() {
+  SkBitmap image = *(new SkBitmap);
```

Comment:
This variable is redundant, we can pass `icon_.AsBitmap()` to `Show(...)`, when icon is empty it will just return an empty bitmap.